### DISCUSSION TITLE
Zephyr USB-RNDIS

### DISCUSSION THREAD

**Discussion by 99003662** [zephyrproject-rtos/zephyr#47865] [Q&A] [answered]
HI team,
            I am following the below link to run the usb- rndis example. I have modified the conf according to the link . however it is not working .  it is not detecting the usb driver itself( i.e in network adaptor it should detect that remote ndis compatible  device).  [ https://github.com/zephyrproject-rtos/zephyr/issues/44573 ].  Do we have any direct example to test this functionality.
 
Can you please help me to figure it out.

Note - Am using windows system , board - STM32 board.

**Comment by enthunilu** [comment, marked as answer]
Hello @99003662 

I also faced the same issue. 

Please try this at your end if it works,
I tried to change rndis_cfg structure available in the below code
[https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/netusb/function_rndis.c]()

as

`
USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_rndis_config rndis_cfg = {
	.iad = {
		.bLength = sizeof(struct usb_association_descriptor),
		.bDescriptorType = USB_DESC_INTERFACE_ASSOC,
		.bFirstInterface = 0,
		.bInterfaceCount = 0x02,
		.bFunctionClass = 0xEF, //USB_BCC_CDC_CONTROL,
		.bFunctionSubClass = 4, //6,
		.bFunctionProtocol = 1, //0,
		.iFunction = 0,
	},
	/* Interface descriptor 0 */
	/* CDC Communication interface */
	.if0 = {
		.bLength = sizeof(struct usb_if_descriptor),
		.bDescriptorType = USB_DESC_INTERFACE,
		.bInterfaceNumber = 0,
		.bAlternateSetting = 0,
		.bNumEndpoints = 1,
		.bInterfaceClass = 0xEF, //USB_BCC_CDC_CONTROL
		.bInterfaceSubClass = 4, // ACM_SUBCLASS,
		.bInterfaceProtocol = 1, // ACM_VENDOR_PROTOCOL
		.iInterface = 0,
	},
`

![image](https://user-images.githubusercontent.com/93653544/179446793-a1b1a63a-699d-4a9b-8e26-52146ab6305b.png)

with above change i tried to use below example and I am able to see device getting detected as RNDIS device 
[https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/zperf]

However, functionality is still not working. So, I am looking at it. Can you also try the same at your end ?

One more thing device is not getting detected for 
[https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/sockets/dumb_http_server]()

as mentioned in [https://github.com/zephyrproject-rtos/zephyr/issues/44573]()

For this you need to enable USB, something like below code along with existing ethernet code. Unless we enable usb, device will not be shown

``	ret = usb_enable(NULL);
	if (ret != 0) {
		printk("usb enable error %d\n", ret);
	}
	(void)net_config_init_app(NULL, "Initializing network");
``

**Comment by finikorg** [comment]
There is a possibility to use USB OS Descriptors like described here: https://static.sched.com/hosted_files/osseu19/6a/WYOG-eng-v15.pdf
